Swan Lake alpha5 release note for runtime changes (#161)
ballerina-platform/ballerina-release: release-notes/swan-lake-alpha5-release-note.md

#### New Features

> ###### Support for configurable variables with records having fields of record types

Reviewer: ```suggestion
###### Support for Configurable Variables with Records Having Fields of Record Types
```

```ballerina
public type Person readonly & record {
     string name;
     int id;
     Address address;
};

public type Address  record {
    string city;
    County country;
};

public type County  record {
    string name;
};

configurable Person person = ?;

```

The  `Config.toml` would be as follows.

```toml
[person]
name = "waruna"
id = 10
address.city="Colombo"
address.country.name="Sri Lanka"

```

> ###### Support for configurable variables with arrays having fields of record types

Reviewer: ```suggestion
###### Support for Configurable Variables with Arrays Having Fields of Record Types
```

```ballerina
configurable Person[] & readonly personArray = ?;

```
The  `Config.toml` would be as follows.

```toml
[[personArray]]
name = "manu"
id = 11
address.city="New York"
address.country.name="USA"

[[personArray]]
name = "hinduja"
id = 12
address.city="London"
address.country.name="UK"
```

> ###### Support for configurable variables with multidimentional arrays

Reviewer: ```suggestion
###### Support for Configurable Variables with Multidimentional Arrays
```

```ballerina
configurable int[][] & readonly int2DArr = ?;
```
The  `Config.toml` would be as follows.

```toml
int2DArr = [[1,2],[3,4]]
```

> ###### Support for optional module name in TOML syntax of configurable variables

Reviewer: ```suggestion
###### Support for Optional Module Name in TOML Syntax of Configurable Variables
```

- operands

**Options**

Included record parameter as the last of the parameter specify options.

```ballerina
public type Person record {
	string name;
	float? score = 0;
};

public function main(*Person person) {
	// Process data here
}
```

```
bal run file.bal -- --name riyafa --score=99.9
```
In the above example, `name` and `score` are options. `riyafa` and `99.9` are arguments of the option.

**Operands**

Other parameters that are not included records specify operands; for these parameters, the position is significant and the name is not.

```ballerina
public type Person record {
	string name;
	float? score = 0;
};
public function main(int efficiency, string character, *Person person) {
	// Process data here
}
```

```
bal run file.bal -- --name riyafa  100 --score=99.9 Good
```

This example, which is the same as above includes `100`, which gets mapped to `efficiency`, and `Good`, which gets mapped to `character` as operands. 

Both operand and option parameters can be of types int, float, decimal, string, array of any of these types and union of any of these types with nil. 
Additionally, option parameters can be of types `boolean`, `boolean[]`, or `boolean?`.

> **Operand arrays**

Reviewer: ```suggestion
**Operand Arrays**
```

>**Note:** If there is an operand parameter of type O[], then it cannot be followed by parameters of type O[], O?, and O x = d. Here O stands for a type that is a subtype of one of string, float, or decimal. An array value is specified by repeatedly specifying the `option` parameter.

If `scores` is an int array, then,

```ballerina
bal run file.bal -- --scores=10 --scores=20 --scores=30
```
This produces the following int array.
```
[10, 20, 30]
```

> **Option boolean parameters**

Reviewer: ```suggestion
**Option Boolean Parameters**
```

#### New Features

> ###### Support for configure variables with records of records, record arrays and multi dimensional array types with Config.toml.

Author: ###### Support for configurable variables with records having fields of record types

> ###### Support for configure variables with records of records, record arrays and multi dimensional array types with Config.toml.

Author: ```suggestion
```

> **Record of records**

Reviewer 2: Shall we say `Record with record field type` or `Record with fields of type record`?

Reviewer: This is too long for a heading. Is it possible to summarize and make it shorter?

E.g., "Support for configurable variables with records and arrays"?

Author: We already have support for records and arrays. These are some complex cases related to records and arrays. Maybe I can have 3 headings.

1) Support for configurable variables with records with fields of type record
2) Support for configurable variables with arrays with fields of type record
2) Support for configurable variables with multidimensional arrays

Reviewer 2: How about `Improved support for composite configurables`

Author: Then we need to consider these as improvements than new features. That should be Ok I think.

Author: BTW i have removed the main heading and broken it into 3 short headings.

Reviewer: +1

> **Record of records**

Author: ```suggestion
###### Support for configurable variables with records having fields of record types
```

```ballerina
public type Person readonly & record {
     string name;
     int id;
     Address address;
};

public type Address  record {
    string city;
    County country;
};

public type County  record {
    string name;
};

configurable Person person = ?;

```

> In  `Config.toml`

Reviewer: ```suggestion
The  `Config.toml` would be as follows.
```

```toml
[person]
name = "waruna"
id = 10
address.city="Colombo"
address.country.name="Sri Lanka"

```

> **Record Arrays**

Author: ```suggestion
###### Support for configurable variables with arrays having fields of record types
```

```ballerina
configurable Person[] & readonly personArray = ?;

```

> In  `Config.toml`

Reviewer: ```suggestion
The  `Config.toml` would be as follows.
```

```toml
[[personArray]]
name = "manu"
id = 11
address.city="New York"
address.country.name="USA"

[[personArray]]
name = "hinduja"
id = 12
address.city="London"
address.country.name="UK"
```

> **Multidimensional Arrays**

Author: ```suggestion
###### Support for configurable variables with multidimentional arrays
```

```ballerina
configurable int[][] & readonly int2DArr = ?;
```

> In  `Config.toml`

Reviewer: ```suggestion
The  `Config.toml` would be as follows.
```

```toml
int2DArr = [[1,2],[3,4]]
```

> ###### Support for optional module name specification in TOML syntax of configurable variables.

Reviewer 2: ```suggestion
###### Support for optional module name in TOML syntax of configurable variables
```

> When providing values for the configurable variables, the  module information can be provided in `Config.toml` according to the following specifications.

Reviewer 2: ```suggestion
When providing values for the configurable variables, the  module information should be provided in `Config.toml` according to the following specifications.
```

> When providing values for the configurable variables, the  module information can be provided in `Config.toml` according to the following specifications.

Reviewer: ```suggestion
When providing values for the configurable variables, the  module information can be provided in the `Config.toml` file according to the following specifications.
```

> When providing values for the configurable variables, the  module information can be provided in `Config.toml` according to the following specifications.

Author: ```suggestion
When providing values for the configurable variables, the module information should be provided in the `Config.toml` file according to the following specifications.
```

* The org-name and module-name are optional for configurable variables defined in the root module of the program.
* The org-name is optional only for configurable variables defined in the root package of the program.

> For example, consider a package with organization name `myOrg` and root module name `main`. 

Reviewer: ```suggestion
For example, consider a package with the organization name as `myOrg` and root module name as `main`. 
```

> In `main.bal`

Reviewer: ```suggestion
The  `main.bal` would be as follows.
```

```ballerina
import main.foo;
import importedOrg/mod;
configurable string mainVar = ?;
public function main() {
// use imported modules
}
```

> In `foo.bal` from module `main.foo`

Reviewer: ```suggestion
The`foo.bal` file of the `main.foo` module will be as follows.
```

```ballerina
configurable string fooVar = ?;
```

> In `mod.bal` which is from another package with organization name `importedOrg` and module name `mod`, 

Reviewer: ```suggestion
In the `mod.bal`, which is from another package with the organization name `importedOrg` and module name `mod`, 
```

```ballerina
configurable string modVar = ?;
```

> The values can be provided in `Config.toml` as below.

Reviewer: ```suggestion
The values can be provided in `Config.toml` as follows.
```

```toml
mainVar = "variable from root module"
[main.foo]
fooVar = "variable from non-root module of the root package"
[importedOrg.mod]
modVar = "variable from non-root package"
```
#### Improvements

> ###### Improved command line argument parsing

Reviewer: ```suggestion
###### Improved Command-Line Argument Parsing
```

> The command line arguments are now parsed into:

Reviewer: ```suggestion
The command-line arguments are now parsed into:
```

> options

Reviewer: ```suggestion
- options
```

> option arguments

Reviewer: ```suggestion
- option arguments
```

> operands

Reviewer: ```suggestion
- operands
```

**Options**

> Included record parameter as the last parameter specify options.

Reviewer: ```suggestion
Included record parameter as the last of the parameter specify options.
```

```ballerina
public type Person record {
	string name;
	float? score = 0;
};

public function main(*Person person) {
	// Process data here
}
```

```
bal run file.bal -- --name riyafa --score=99.9
```

> In the above example `name` and `score` are options. `riyafa` and `99.9` are option arguments.

Reviewer: ```suggestion
In the above example, `name` and `score` are options. `riyafa` and `99.9` are arguments of the option.
```

**Operands**

Other parameters that are not included records specify operands; for these parameters, the position is significant and the name is not.

```ballerina
public type Person record {
	string name;
	float? score = 0;
};
public function main(int efficiency, string character, *Person person) {
	// Process data here
}
```

```
bal run file.bal -- --name riyafa  100 --score=99.9 Good
```

> This example which is the same as above includes `100` which gets mapped to `efficiency` and `Good` which gets mapped to `character` as operands. 

Reviewer: ```suggestion
This example, which is the same as above includes `100`, which gets mapped to `efficiency`, and `Good`, which gets mapped to `character` as operands. 
```

Both operand and option parameters can be of types int, float, decimal, string, array of any of these types and union of any of these types with nil. 

> Additionally option parameters can be of types `boolean`, `boolean[]` or `boolean?`.

Reviewer: ```suggestion
Additionally, option parameters can be of types `boolean`, `boolean[]`, or `boolean?`.
```

**Operand arrays**

> Note that if there is an operand parameter of type O[], then it cannot be followed by parameters of type O[], O? and O x = d. Here O stands for a type that is a subtype of one of string,float or decimal. An array value is specified by repeatedly specifying the option parameter.

Reviewer: ```suggestion
>**Note:** If there is an operand parameter of type O[], then it cannot be followed by parameters of type O[], O?, and O x = d. Here O stands for a type that is a subtype of one of string, float, or decimal. An array value is specified by repeatedly specifying the `option` parameter.
```

> If `scores` is an int array then,

Reviewer: ```suggestion
If `scores` is an int array, then,
```

```ballerina
bal run file.bal -- --scores=10 --scores=20 --scores=30
```

> This produces the following int array

Reviewer: ```suggestion
This produces the following int array.
```

```
[10, 20, 30]
```

**Option boolean parameters**

> When there’s an option of `boolean`, `boolean[]` or `boolean?` type it does not take an option argument. The presence of the option is considered to be `true` and the absence of it is considered to be false. 

Reviewer: ```suggestion
When there’s an option of `boolean`, `boolean[]`, or `boolean?` type, it does not take an `option` argument. The presence of the option is considered to be `true` and the absence of it is considered to be false. 
```

> In the following example suppose `results` is a boolean array.

Reviewer: ```suggestion
In the following example, suppose `results` is a boolean array.
```

```
bal run file.bal -- --results --results --results
```

> This produces the following boolean array

Reviewer: ```suggestion
This produces the following boolean array.
```
